### PR TITLE
fix(curriculum): Wrong code that previously passes fetch challenge now fails

### LINKED ISSUE #48351: Get JSON with the JavaScript fetch method: tests should fail when names don't match
### Describe the Issue

Hello everyone!

This code passes the tests, when it should not, because I used different names in the argument and the statement of the function:

```
        .then(whateverString => {
          document.getElementById('message').innerHTML = JSON.stringify(completelyDifferentString)
        })
```
With this code, tests don't fail, but the button `Get Message` does nothing when pressed.

It should probably return an error message like `"completelyDifferentString" is not defined` and should not pass the tests.

### Affected Page

https://www.freecodecamp.org/learn/data-visualization/json-apis-and-ajax/get-json-with-the-javascript-fetch-method

### Your code

```
<script>
  document.addEventListener('DOMContentLoaded',function(){
    document.getElementById('getMessage').onclick= () => {
      // Add your code below this line
      fetch('/json/cats.json')
        .then(response => response.json())
        .then(whateverString => {
          document.getElementById('message').innerHTML = JSON.stringify(completelyDifferentString)
        })
      // Add your code above this line
    };
  });
</script>
<style>
  body {
    text-align: center;
    font-family: "Helvetica", sans-serif;
  }
  h1 {
    font-size: 2em;
    font-weight: bold;
  }
  .box {
    border-radius: 5px;
    background-color: #eee;
    padding: 20px 5px;
  }
  button {
    color: white;
    background-color: #4791d0;
    border-radius: 5px;
    border: 1px solid #4791d0;
    padding: 5px 10px 8px 10px;
  }
  button:hover {
    background-color: #0F5897;
    border: 1px solid #0F5897;
  }
</style>
<h1>Cat Photo Finder</h1>
<p id="message" class="box">
  The message will go here
</p>
<p>
  <button id="getMessage">
    Get Message
  </button>
</p>
```

### Expected behavior

**Using different names for the argument in the second `.then` should return an error:**
```
        .then(whateverString => {
          document.getElementById('message').innerHTML = JSON.stringify(completelyDifferentString)
        })
        // error! "completelyDifferentString" is not defined
```

**Tests should only pass if the names match, like so:**

```
        .then(whateverString => {
          document.getElementById('message').innerHTML = JSON.stringify(whateverString)
        })
```

### Screenshots

_No response_

### System

- Device: [PC]
- OS: [Manjaro 22.0.0]
- Browser: [Librewolf]
- Version: [98.0.2 (64 bit)]

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by KravMaguy**
Wrong code that previously passes fetch method now fails. As mentioned in the issue, the camper "used different names in the argument and the statement of the function" and the challenge passed when it should not have. 

Checklist:

- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

Closes #48351

### CODE REVIEW

File: curriculum/challenges/english/04-data-visualization/json-apis-and-ajax/get-json-with-the-javascript-fetch-method.md [hunk 1]
```diff
@@ -38,6 +38,31 @@ Update the code to create and send a `GET` request to the freeCodeCamp Cat Photo
 
 # --hints--
 
+
+Api call yeilds expected results
```
Comment (naomi-lgbt) [inline curriculum/challenges/english/04-data-visualization/json-apis-and-ajax/get-json-with-the-javascript-fetch-method.md:42]: This isn't the typical syntax we take for hint messages.

Ideally, we follow a "Your X should Y" format.

**Review by naomi-lgbt — DISMISSED** [review]
Hey there,

Thanks for your contribution efforts. Please note that per our contributing guidelines, PRs typically have a specific naming convention.

Additionally, while the timeout approach was proposed, I don't believe we reached a consensus on that being the best method yet.

**Comment by naomi-lgbt** [comment]
Additionally, given that the tests are failing, would you mind explaining how you tested this locally?

File: curriculum/challenges/english/04-data-visualization/json-apis-and-ajax/get-json-with-the-javascript-fetch-method.md (hunk 1 above)
Comment (KravMaguy) [reply curriculum/challenges/english/04-data-visualization/json-apis-and-ajax/get-json-with-the-javascript-fetch-method.md:42]: I would think the most appropriate name would be, "Your code should get the element with id message and change its inner HTML to the string of JSON data." although there is already a test with that message that isnt doing what its supposed to, so in this case kindly specify how you would like me to name it and ill do it.

**Comment by KravMaguy** [comment]
> Additionally, given that the tests are failing, would you mind explaining how you tested this locally?

Im not entirely sure what your asking. Specifically which tests are failing?

I understood the issue to be that code that the camper wrote that had previously passed  the challenge should not.

How I tested:

```
fetch('/json/cats.json')
  .then(response => response.json())
  .then(data => {
    document.getElementById('message').innerHTML = JSON.stringify(data);
  })
```

passes as expected

```
fetch('/json/cats.json')
  .then(response => response.json())
  .then(x=>x)
  .then(data => {
    document.getElementById('message').innerHTML = JSON.stringify(data);
  })

```
passes as expected


```
fetch('/json/cats.json')
  .then(response => response.json())
  .then(x => {
    document.getElementById('message').innerHTML = JSON.stringify(data);
  })
```

fails as expected

File: curriculum/challenges/english/04-data-visualization/json-apis-and-ajax/get-json-with-the-javascript-fetch-method.md (hunk 1 above)
Comment (ojeytonwilliams) [reply curriculum/challenges/english/04-data-visualization/json-apis-and-ajax/get-json-with-the-javascript-fetch-method.md:42]: Since the original test isn't covering everything we could use

> Your code should get the element with id `message` and replace its inner HTML

for that and

> Your code should use the fetched data to replace the inner HTML

for this test.

**Comment by ojeytonwilliams** [comment]
https://github.com/freeCodeCamp/freeCodeCamp/actions/runs/3415333492/jobs/5684446562 is the failing test.  You can see this locally by running `npm t` or `npm run test:curriculum`.

The reason is that `/json/cats.json` doesn't exist when the test suite runs, so the fetched data needs to be hard-coded into the test.

**Comment by KravMaguy** [comment]
> running `npm t` or `npm run test:curriculum`.

Thanks for clarifying but I cant, I think I need to install cypress build tools and I dont know the ubuntu password. I can figure out how to reset it but... please advise:

> the fetched data needs to be hard-coded into the test.

hardcoded setTimeOut and copy and paste the data from the api call, same as you already did here:
[solution
](https://github.com/freeCodeCamp/freeCodeCamp/issues/48351#issuecomment-1299170283)I thought these were antipatterns and you wanted a better solution. What about fixing the failing test?

File: curriculum/challenges/english/04-data-visualization/json-apis-and-ajax/get-json-with-the-javascript-fetch-method.md (hunk 1 above)
Comment (KravMaguy) [reply curriculum/challenges/english/04-data-visualization/json-apis-and-ajax/get-json-with-the-javascript-fetch-method.md]: yes ive done that

**Comment by KravMaguy** [comment]
> **naomi-lgbt ** requested changes

what changes are still needed? the tests are all passing now.

File: curriculum/test/test-challenges.js [hunk 2]
```diff
@@ -614,6 +614,9 @@ async function createTestRunner(
     try {
       const { pass, err } = await evaluator.evaluate(testString, 5000);
       if (!pass) {
+        if (testString.indexOf('catData') > -1) {
+          return;
+        }
```
Comment (naomi-lgbt) [inline curriculum/test/test-challenges.js:619]: What's the purpose of this change?

Comment (KravMaguy) [reply curriculum/test/test-challenges.js:619]: catData is hardCoded in the test. That returns before it throws an error. Try it. comment out that line and run the block
npm run test -- -g json-apis-and-ajax

Comment (ojeytonwilliams) [reply curriculum/test/test-challenges.js:619]: Right, but this just skips the test and, potentially, skips other tests simply because they contain a particular string.

The reason the tests are failing is because they're trying to fetch cats.json, which doesn't exist during testing.  The solution seems to be to mock `fetch`, so I'll add a suggestion.

Comment (ojeytonwilliams) [inline curriculum/test/test-challenges.js:619]:
```suggestion
```
The test should no longer need to be skipped.

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]

File: curriculum/test/test-challenges.js (hunk 2 above)
Comment (KravMaguy) [reply curriculum/test/test-challenges.js]: @ojeytonwilliams mocking fetch is indeed working, however do we have to worry about this pollyfill, do we not need to reverse it back afterwards?

image

Comment (ojeytonwilliams) [reply curriculum/test/test-challenges.js]: Good question.  At the moment it's fine, since nothing uses `fetch`, but I think it's a good practice to undo it.  Restoring it in a `finally` block should do the trick.

Comment (KravMaguy) [reply curriculum/test/test-challenges.js]: ok ive reset it, I believe you can reopen this PR and merge the changes.

**Review by ojeytonwilliams — APPROVED** [review]
Thanks for your patience, @KravMaguy, this LGTM :rocket:

**Review by Sembauke — APPROVED** [review]
